File: sdk/python/tests/cucumber/steps/steps.py
"""Thin step definitions for HTTP client sanity Gherkin tests (Python/behave).

Execute the SDK request and write actual JSON. All assertion/normalization
logic is delegated to the shared judge_scenario.js (single source of truth).
"""
import asyncio
import base64
import json
import os
import subprocess
import time

from behave import given, when, then

from payments.http_client import execute, HttpRequest, create_client, merge_http_config, DEFAULT_HTTP_CONFIG
from payments.generated import sdk_config_pb2

LANG = 'python'
JUDGE = os.path.abspath(os.path.join(
    os.path.dirname(__file__), '..', '..', '..', '..', 'tests', 'client_sanity', 'judge_scenario.js'))


# ── Given ────────────────────────────────────────────────────────

@given('the echo server is running on port {port:d}')
def step_echo_server(context, port):
    pass  # started externally


@given('a "{method}" request to "{url}"')
def step_request(context, method, url):
    context.method = method
    context.url = url


@given('header "{name}" is "{value}"')
def step_header(context, name, value):
    context.headers[name] = value


@given("body is '{body}'")
def step_body_single(context, body):
    context.body = body.replace('\\r\\n', '\r\n').replace('\\n', '\n')


@given('body is "{body}"')
def step_body_double(context, body):
    context.body = body.replace('\\r\\n', '\r\n').replace('\\n', '\n')


@given('a response timeout of {ms:d} ms')
def step_timeout(context, ms):
    context.response_timeout_ms = ms


@given('the proxy is "{url}"')
def step_proxy(context, url):
    context.proxy_url = url


# ── When (thin: execute + write actual JSON) ─────────────────────

@when('the request is sent as scenario "{scenario_id}"')
def step_execute(context, scenario_id):
    context.scenario_id = scenario_id
    context.source_id = f'{LANG}_{scenario_id}'

    actual_file = os.path.join(context.artifacts_dir, f'actual_{context.source_id}.json')
    capture_file = os.path.join(context.artifacts_dir, f'capture_{context.source_id}.json')
    for f in [actual_file, capture_file]:
        if os.path.exists(f):
            os.unlink(f)

    headers = dict(context.headers)
    headers['x-source'] = context.source_id
    headers['x-scenario-id'] = scenario_id

    body = context.body
    if isinstance(body, str) and body.startswith('base64:'):
        body = base64.b64decode(body[7:])
    elif isinstance(body, str):
        body = body.encode('utf-8')

    client_config = None
    if context.proxy_url:
        client_config = sdk_config_pb2.HttpConfig(
            proxy=sdk_config_pb2.ProxyOptions(http_url=context.proxy_url))

    try:
        client = create_client(client_config)
    except Exception as e:
        code = getattr(e, 'error_code', None) or getattr(e, 'errorCode', None) or 'UNKNOWN_ERROR'
        _write_json(actual_file, {'error': {'code': str(code), 'message': str(e)}})
        return

    override = None
    if context.response_timeout_ms is not None:
        override = sdk_config_pb2.HttpConfig(response_timeout_ms=context.response_timeout_ms)

    resolved_ms = None
    if override:
        base = client_config if client_config else DEFAULT_HTTP_CONFIG
        merged = merge_http_config(base, override)
        resolved_ms = (merged.total_timeout_ms, merged.connect_timeout_ms, merged.response_timeout_ms)

    request = HttpRequest(url=context.url, method=context.method, headers=headers, body=body)
    output = {}

    try:
        resp = asyncio.get_event_loop().run_until_complete(execute(request, client, resolved_ms))
        ct = (resp.headers.get('content-type', '') or '').lower()
        body_str = (base64.b64encode(resp.body).decode('utf-8')
                    if 'application/octet-stream' in ct
                    else resp.body.decode('utf-8', errors='replace'))
        output['response'] = {'statusCode': resp.status_code, 'headers': resp.headers, 'body': body_str}
    except Exception as e:
        code = getattr(e, 'error_code', None) or getattr(e, 'errorCode', None) or 'UNKNOWN_ERROR'
        output['error'] = {'code': str(code), 'message': str(e)}

    _write_json(actual_file, output)
    time.sleep(0.2)  # wait for echo server capture


# ── Then (delegate ALL assertions to the shared judge) ───────────

@then('the response status should be {status:d}')
def step_status(context, status):
    pass  # validated by judge


@then("the response body should be '{expected}'")
def step_body_check_single(context, expected):
    pass  # validated by judge


@then('the response body should be \'{expected}\'')
def step_body_check_escaped(context, expected):
    pass  # validated by judge


@then('the response header "{name}" should be "{value}"')
def step_header_check(context, name, value):
    pass  # validated by judge


@then('the response should have multi-value header "{name}" with values "{values_str}"')
def step_multi_header_check(context, name, values_str):
    pass  # validated by judge


@then('the SDK should return error "{expected_code}"')
def step_error_check(context, expected_code):
    _run_judge(context)


@then('the server should have received the correct request')
def step_capture_check(context):
    _run_judge(context)


# ── Helpers ──────────────────────────────────────────────────────

def _write_json(filepath, data):
    with open(filepath, 'w') as f:
        json.dump(data, f, indent=2)


def _run_judge(context):
    if getattr(context, '_judged', False):
        return
    context._judged = True
    result = subprocess.run(
        ['node', JUDGE, LANG, context.scenario_id],
        capture_output=True, text=True)
    if result.returncode != 0:
        msg = f'Judge FAILED for {context.scenario_id}'
        try:
            msg = json.loads(result.stdout)['message']
        except Exception:
            pass
        raise AssertionError(msg)

File: sdk/javascript/tests/cucumber/steps.ts
import { Given, When, Then, Before } from '@cucumber/cucumber';
import { SanityWorld } from './world';
import { execute, createDispatcher } from '../../src/http_client';
import { execFileSync } from 'child_process';
import * as fs from 'fs';
import * as path from 'path';

const LANG = 'node';
const JUDGE = path.resolve(__dirname, '../../../tests/client_sanity/judge_scenario.js');
const MANIFEST_PATH = path.resolve(__dirname, '../../../tests/client_sanity/manifest.json');

// Build title → id lookup once at import time.
const manifest = JSON.parse(fs.readFileSync(MANIFEST_PATH, 'utf8'));
const titleToId: Record<string, string> = {};
for (const s of manifest.scenarios) {
  if (s.title) titleToId[s.title] = s.id;
}

Before(function (this: SanityWorld, { pickle }) {
  if (pickle.tags.some(t => t.name === '@skip_node')) return 'skipped';
  // Resolve scenario ID from the Gherkin scenario title.
  this.scenarioId = titleToId[pickle.name] || '';
  this.sourceId = `${LANG}_${this.scenarioId}`;
});

// ── Given ───────────────────────────────────────────────────────

Given('the echo server is running on port {int}', function (this: SanityWorld, port: number) {
  this.baseUrl = `http://localhost:${port}`;
});

Given('a {string} request to {string}', function (this: SanityWorld, method: string, url: string) {
  this.method = method;
  this.url = url;
});

Given('query parameter {string} is {string}', function (this: SanityWorld, name: string, value: string) {
  this.queryParams.push([name, value]);
});

Given('header {string} is {string}', function (this: SanityWorld, name: string, value: string) {
  this.headers[name] = value;
});

Given('body is {string}', function (this: SanityWorld, body: string) {
  this.body = body.replace(/\\r\\n/g, '\r\n').replace(/\\n/g, '\n');
});

Given('body is:', function (this: SanityWorld, docString: string) {
  // Doc strings use \n; multipart bodies need \r\n line endings and a trailing \r\n.
  const ct = (this.headers['Content-Type'] || '').toLowerCase();
  if (ct.includes('multipart/')) {
    this.body = docString.replace(/\n/g, '\r\n') + '\r\n';
  } else {
    this.body = docString;
  }
});

Given('a response timeout of {int} ms', function (this: SanityWorld, ms: number) {
  this.responseTimeoutMs = ms;
});

Given('the proxy is {string}', function (this: SanityWorld, url: string) {
  this.proxyUrl = url;
});

// ── When (thin: execute + write actual JSON) ────────────────────

When('the request is sent', async function (this: SanityWorld) {
  if (!this.scenarioId) throw new Error('Could not resolve scenario ID from Gherkin title');

  const actualFile = path.join(this.getArtifactsDir(), `actual_${this.sourceId}.json`);
  const captureFile = this.getCaptureFile();
  if (fs.existsSync(captureFile)) fs.unlinkSync(captureFile);
  if (fs.existsSync(actualFile)) fs.unlinkSync(actualFile);

  const fullUrl = this.resolveUrl();

  // Build request
  const request: any = {
    method: this.method,
    url: fullUrl,
    headers: { ...this.headers, 'x-source': this.sourceId, 'x-scenario-id': this.scenarioId },
    body: this.body,
  };
  if (typeof request.body === 'string' && request.body.startsWith('base64:')) {
    request.body = Uint8Array.from(Buffer.from(request.body.replace('base64:', ''), 'base64'));
  }

  const opts: any = {};
  if (this.responseTimeoutMs != null) opts.responseTimeoutMs = this.responseTimeoutMs;

  const dispatcherConfig: any = { ...opts };
  if (this.proxyUrl) dispatcherConfig.proxy = { httpUrl: this.proxyUrl };

  let dispatcher: any;
  try {
    dispatcher = createDispatcher(dispatcherConfig);
  } catch (e: any) {
    const code = e?.errorCode ?? (typeof e?.code === 'string' ? e.code : 'UNKNOWN_ERROR');
    fs.writeFileSync(actualFile, JSON.stringify({ error: { code, message: e?.message || String(e) } }, null, 2));
    return;
  }

  let output: any = {};
  try {
    const resp = await execute(request, opts, dispatcher);
    const ct = (resp.headers['content-type'] || '').toLowerCase();
    const bodyStr = ct.includes('application/octet-stream')
      ? Buffer.from(resp.body).toString('base64')
      : new TextDecoder().decode(resp.body);
    output.response = { statusCode: resp.statusCode, headers: resp.headers, body: bodyStr };
  } catch (e: any) {
    const code = e?.errorCode ?? (typeof e?.code === 'string' ? e.code : 'UNKNOWN_ERROR');
    output.error = { code, message: e?.message || String(e) };
  }

  fs.writeFileSync(actualFile, JSON.stringify(output, null, 2));
  await new Promise(r => setTimeout(r, 200)); // wait for echo server capture
});

// ── Then (delegate ALL assertions to the shared judge) ──────────

Then('the response status should be {int}', function () { /* validated by judge */ });
Then('the response body should be {string}', function () { /* validated by judge */ });
Then('the response header {string} should be {string}', function () { /* validated by judge */ });
Then('the response should have multi-value header {string} with values {string}', function () { /* validated by judge */ });

Then('the SDK should return error {string}', function (this: SanityWorld) {
  runJudge(this);
});

Then('the server should have received the correct request', function (this: SanityWorld) {
  runJudge(this);
});

function runJudge(world: SanityWorld) {
  if (world.judged) return; // only run once per scenario
  world.judged = true;
  try {
    execFileSync('node', [JUDGE, LANG, world.scenarioId], { stdio: ['pipe', 'pipe', 'inherit'] });
  } catch (e: any) {
    const stdout = e.stdout?.toString() || '';
    let msg = `Judge FAILED for ${world.scenarioId}`;
    try { msg = JSON.parse(stdout).message; } catch {}
    throw new Error(msg);
  }
}
